File: Fort.ind UWP/Services/LiveTileService.vb
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Text
Imports Windows.UI.Notifications
Imports Windows.Data.Xml.Dom

''' <summary>
''' Service for managing Live Tile updates with news and notifications
''' </summary>
Public Class LiveTileService

    ''' <summary>
    ''' Updates the Live Tile with the latest news
    ''' </summary>
    Public Shared Sub UpdateTileWithNews(title As String, message As String, Optional branding As String = "name", Optional animationType As TileAnimation = TileAnimation.FadeIn)
        Try
            ' Create the tile notification content
            Dim tileXml = CreateTileXml(title, message, branding, animationType)

            ' Create and send the notification
            Dim tileNotification As New TileNotification(tileXml)
            TileUpdateManager.CreateTileUpdaterForApplication().Update(tileNotification)
        Catch ex As Exception
            Debug.WriteLine($"LiveTileService: UpdateTileWithNews failed – {ex.GetType().Name}: {ex.Message}")
        End Try
    End Sub

    ''' <summary>
    ''' Updates the Live Tile with multiple news items that cycle
    ''' </summary>
    Public Shared Sub UpdateTileWithMultipleNews(newsItems As List(Of NewsItem))
        If newsItems Is Nothing OrElse newsItems.Count = 0 Then Return

        Try
            ' Enable notification queue to show multiple tiles
            Dim tileUpdater = TileUpdateManager.CreateTileUpdaterForApplication()
            tileUpdater.EnableNotificationQueue(True)

            ' Clear existing notifications
            tileUpdater.Clear()

            ' Animation types to cycle through
            Dim animations As TileAnimation() = {
                TileAnimation.FadeIn,
                TileAnimation.SlideUp,
                TileAnimation.SlideDown,
                TileAnimation.SlideLeft,
                TileAnimation.SlideRight
            }

            ' Add each news item (max 5 in queue)
            For i = 0 To Math.Min(newsItems.Count - 1, 4)
                Dim item = newsItems(i)
                If item Is Nothing Then Continue For

                Dim animation = animations(i Mod animations.Length)
                Dim tileXml = CreateTileXml(item.Title, item.Message, "name", animation)
                Dim tileNotification As New TileNotification(tileXml)
                tileNotification.Tag = If(String.IsNullOrWhiteSpace(item.Tag), $"news{i}", item.Tag)
                tileUpdater.Update(tileNotification)
            Next
        Catch ex As Exception
            Debug.WriteLine($"LiveTileService: UpdateTileWithMultipleNews failed – {ex.GetType().Name}: {ex.Message}")
        End Try
    End Sub

    ''' <summary>
    ''' Creates the tile XML for different tile sizes with animations
    ''' </summary>
    Private Shared Function CreateTileXml(title As String, message As String, branding As String, Optional animation As TileAnimation = TileAnimation.FadeIn) As XmlDocument
        Dim textStyleAttr = GetTextStyleAttribute(animation)
        Dim safeBranding = EscapeXml(branding)
        Dim safeTitle = EscapeXml(title)
        Dim safeMessage = EscapeXml(message)
        Dim smallTileText = EscapeXml(GetTileMonogram(title, branding))

        ' Adaptive tile template supporting all sizes with validated content
        Dim tileXmlString = $"
<tile>
    <visual branding=""{safeBranding}"" displayName=""Fort.ind"">
        <binding template=""TileSmall"" hint-textStacking=""center"">
            <text hint-style=""caption"" hint-align=""center"">{smallTileText}</text>
        </binding>

        <!-- Medium Tile (150x150) -->
        <binding template=""TileMedium"">
            <group>
                <subgroup>
                    <text {textStyleAttr} hint-wrap=""true"">{safeTitle}</text>
                    <text hint-style=""captionSubtle"" hint-wrap=""true"" hint-maxLines=""3"">{safeMessage}</text>
                </subgroup>
            </group>
        </binding>

        <!-- Wide Tile (310x150) -->
        <binding template=""TileWide"">
            <group>
                <subgroup>
                    <text {textStyleAttr}>{safeTitle}</text>
                    <text hint-style=""body"" hint-wrap=""true"" hint-maxLines=""2"">{safeMessage}</text>
                </subgroup>
            </group>
        </binding>

        <!-- Large Tile (310x310) -->
        <binding template=""TileLarge"" hint-textStacking=""center"">
            <group>
                <subgroup>
                    <text {textStyleAttr} hint-align=""center"">{safeTitle}</text>
                </subgroup>
            </group>
            <text hint-style=""body"" hint-wrap=""true"" hint-maxLines=""6"" hint-align=""center"">{safeMessage}</text>
            <text hint-style=""captionSubtle"" hint-align=""center"">Fort.ind Desktop</text>
        </binding>

    </visual>
</tile>"

        Dim tileXml As New XmlDocument()
        tileXml.LoadXml(tileXmlString)
        Return tileXml
    End Function

    ''' <summary>
    ''' Gets the animation attribute string for the tile
    ''' </summary>
    Private Shared Function GetTextStyleAttribute(animation As TileAnimation) As String
        Select Case animation
            Case TileAnimation.FadeIn
                Return "hint-style=""captionSubtle"""
            Case TileAnimation.SlideUp
                Return "hint-style=""base"""
            Case TileAnimation.SlideDown
                Return "hint-style=""body"""
            Case TileAnimation.SlideLeft
                Return "hint-style=""bodySubtle"""
            Case TileAnimation.SlideRight
                Return "hint-style=""subtitle"""
            Case Else
                Return ""
        End Select
    End Function

    ''' <summary>
    ''' Escapes special XML characters
    ''' </summary>
    Private Shared Function EscapeXml(text As String) As String
        If String.IsNullOrEmpty(text) Then Return ""

        Dim sanitized As New StringBuilder(text.Length)
        For Each ch As Char In text
            If (ch = vbTab) OrElse (ch = vbLf) OrElse (ch = vbCr) OrElse
               (ch >= ChrW(&H20) AndAlso ch <= ChrW(&HD7FF)) OrElse
               (ch >= ChrW(&HE000) AndAlso ch <= ChrW(&HFFFD)) Then
                sanitized.Append(ch)
            End If
        Next

        Return sanitized.ToString().Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("""", "&quot;").Replace("'", "&apos;")
    End Function

    ''' <summary>
    ''' Updates the badge on the tile (shows a number or glyph)
    ''' </summary>
    Public Shared Sub UpdateBadge(count As Integer)
        Try
            If count <= 0 Then
                ClearBadge()
                Return
            End If

            Dim clampedCount = Math.Min(count, 99)
            Dim badgeXml = $"<badge value=""{clampedCount}""/>"
            Dim badgeDoc As New XmlDocument()
            badgeDoc.LoadXml(badgeXml)

            Dim badgeNotification As New BadgeNotification(badgeDoc)
            BadgeUpdateManager.CreateBadgeUpdaterForApplication().Update(badgeNotification)
        Catch ex As Exception
            Debug.WriteLine($"LiveTileService: UpdateBadge failed – {ex.GetType().Name}: {ex.Message}")
        End Try
    End Sub

    ''' <summary>
    ''' Updates the badge with a glyph (icon)
    ''' </summary>
    Public Shared Sub UpdateBadgeGlyph(glyph As String)
        If String.IsNullOrWhiteSpace(glyph) Then Return

        Try
            Dim normalizedGlyph = glyph.Trim()
            If Not IsSupportedBadgeGlyph(normalizedGlyph) Then
                Debug.WriteLine($"LiveTileService: UpdateBadgeGlyph skipped unsupported glyph '{normalizedGlyph}'.")
                Return
            End If

            ' Available glyphs: none, activity, alarm, alert, attention, available, away, busy,
            ' error, newMessage, paused, playing, unavailable
            Dim badgeXml = $"<badge value=""{normalizedGlyph}""/>"
            Dim badgeDoc As New XmlDocument()
            badgeDoc.LoadXml(badgeXml)

            Dim badgeNotification As New BadgeNotification(badgeDoc)
            BadgeUpdateManager.CreateBadgeUpdaterForApplication().Update(badgeNotification)
        Catch ex As Exception
            Debug.WriteLine($"LiveTileService: UpdateBadgeGlyph failed – {ex.GetType().Name}: {ex.Message}")
        End Try
    End Sub

    ''' <summary>
    ''' Shows a Windows toast notification with a title and message.
    ''' Returns False (and writes to Debug output) if notifications are blocked or an error occurs.
    ''' </summary>
    Public Shared Function SendToast(title As String, message As String) As Boolean
        Try
            Dim notifier = ToastNotificationManager.CreateToastNotifier()
            If notifier.Setting <> NotificationSetting.Enabled Then
                Debug.WriteLine($"LiveTileService: Toast suppressed – NotificationSetting is {notifier.Setting}. " &
                                "Enable notifications for this app in Windows Settings > System > Notifications.")
                Return False
            End If

            Dim toastXml As New XmlDocument()
            toastXml.LoadXml(
                $"<toast><visual><binding template=""ToastGeneric"">" &
                $"<text>{EscapeXml(title)}</text>" &
                $"<text>{EscapeXml(message)}</text>" &
                "</binding></visual></toast>")
            Dim toast As New ToastNotification(toastXml)
            notifier.Show(toast)
            Return True
        Catch ex As Exception
            Debug.WriteLine($"LiveTileService: SendToast failed – {ex.GetType().Name}: {ex.Message}")
            Return False
        End Try
    End Function

    ''' <summary>
    ''' Clears the Live Tile back to default
    ''' </summary>
    Public Shared Sub ClearTile()
        TileUpdateManager.CreateTileUpdaterForApplication().Clear()
    End Sub

    ''' <summary>
    ''' Clears the badge
    ''' </summary>
    Public Shared Sub ClearBadge()
        BadgeUpdateManager.CreateBadgeUpdaterForApplication().Clear()
    End Sub

    Private Shared Function GetTileMonogram(primaryText As String, fallbackText As String) As String
        Dim source = If(String.IsNullOrWhiteSpace(primaryText), fallbackText, primaryText)
        If String.IsNullOrWhiteSpace(source) Then Return "FI"

        Dim trimmed = source.Trim()
        If trimmed.Length <= 2 Then Return trimmed.ToUpperInvariant()
        Return trimmed.Substring(0, 2).ToUpperInvariant()
    End Function

    Private Shared Function IsSupportedBadgeGlyph(glyph As String) As Boolean
        Select Case glyph.ToLowerInvariant()
            Case "none", "activity", "alarm", "alert", "attention", "available", "away", "busy", "error", "newmessage", "paused", "playing", "unavailable"
                Return True
            Case Else
                Return False
        End Select
    End Function

End Class

''' <summary>
''' Tile animation types
''' </summary>
Public Enum TileAnimation
    FadeIn
    SlideUp
    SlideDown
    SlideLeft
    SlideRight
End Enum

''' <summary>
''' Represents a news item for the Live Tile
''' </summary>
Public Class NewsItem
    Public Property Title As String
    Public Property Message As String
    Public Property Tag As String
    Public Property Timestamp As DateTime

    Public Sub New(title As String, message As String, Optional tag As String = Nothing)
        Me.Title = title
        Me.Message = message
        Me.Tag = tag
        Me.Timestamp = DateTime.Now
    End Sub
End Class

File: Fort.ind UWP/Services/LiveTileService.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using Windows.UI.Notifications;
using Windows.Data.Xml.Dom;
using Microsoft.Toolkit.Uwp.Notifications;

namespace Fort.ind_UWP
{
    public class LiveTileService
    {
        private const string DefaultMonogram = "FI";

        public const string NewContentBadgeGlyph = "newMessage";

        public static void UpdateTileWithNews(string title, string message, string branding = "name")
        {
            try
            {
                var tileXml = CreateTileXml(title, message, branding);

                TileNotification tileNotification = new TileNotification(tileXml);
                TileUpdateManager.CreateTileUpdaterForApplication().Update(tileNotification);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: UpdateTileWithNews failed – {ex.GetType().Name}: {ex.Message}");
            }
        }

        public static void UpdateTileWithMultipleNews(List<NewsItem> newsItems)
        {
            if (newsItems == null || newsItems.Count == 0) return;

            try
            {
                var tileUpdater = TileUpdateManager.CreateTileUpdaterForApplication();
                tileUpdater.EnableNotificationQueue(true);

                tileUpdater.Clear();

                for (int i = 0; i <= Math.Min(newsItems.Count - 1, 4); i++)
                {
                    var item = newsItems[i];
                    if (item == null) continue;

                    var tileXml = CreateTileXml(item.Title, item.Message, "name");
                    TileNotification tileNotification = new TileNotification(tileXml);
                    tileNotification.Tag = string.IsNullOrWhiteSpace(item.Tag) ? $"news{i}" : item.Tag;
                    tileUpdater.Update(tileNotification);
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: UpdateTileWithMultipleNews failed – {ex.GetType().Name}: {ex.Message}");
            }
        }

        private const AdaptiveTextStyle TitleTextStyle = AdaptiveTextStyle.Base;

        private static XmlDocument CreateTileXml(string title, string message, string branding)
        {
            var titleStyle = TitleTextStyle;
            var safeBranding = ParseBranding(branding);
            var safeTitle = SanitizeText(title);
            var safeMessage = SanitizeText(message);
            var smallTileText = SanitizeText(GetTileMonogram(title));

            TileContent content = new TileContent();
            content.Visual = new TileVisual();
            content.Visual.DisplayName = LocalizedStrings.Get("TileDisplayName");
            content.Visual.Branding = safeBranding;

            AdaptiveText smallText = new AdaptiveText();
            smallText.Text = smallTileText;
            smallText.HintStyle = AdaptiveTextStyle.Caption;
            smallText.HintAlign = AdaptiveTextAlign.Center;

            TileBindingContentAdaptive smallContent = new TileBindingContentAdaptive();
            smallContent.TextStacking = TileTextStacking.Center;
            smallContent.Children.Add(smallText);

            content.Visual.TileSmall = new TileBinding();
            content.Visual.TileSmall.Content = smallContent;

            content.Visual.TileMedium = new TileBinding();
            content.Visual.TileMedium.Content = BuildTitleMessageGroup(safeTitle, titleStyle, true, safeMessage, AdaptiveTextStyle.CaptionSubtle, 3);

            content.Visual.TileWide = new TileBinding();
            content.Visual.TileWide.Content = BuildTitleMessageGroup(safeTitle, titleStyle, false, safeMessage, AdaptiveTextStyle.Body, 2);

            TileBindingContentAdaptive largeContent = new TileBindingContentAdaptive();
            largeContent.TextStacking = TileTextStacking.Center;

            AdaptiveText largeTitleText = new AdaptiveText();
            largeTitleText.Text = safeTitle;
            largeTitleText.HintStyle = titleStyle;
            largeTitleText.HintAlign = AdaptiveTextAlign.Center;

            AdaptiveSubgroup largeSubgroup = new AdaptiveSubgroup();
            largeSubgroup.Children.Add(largeTitleText);

            AdaptiveGroup largeGroup = new AdaptiveGroup();
            largeGroup.Children.Add(largeSubgroup);

            AdaptiveText largeMessageText = new AdaptiveText();
            largeMessageText.Text = safeMessage;
            largeMessageText.HintStyle = AdaptiveTextStyle.Body;
            largeMessageText.HintWrap = true;
            largeMessageText.HintMaxLines = 6;
            largeMessageText.HintAlign = AdaptiveTextAlign.Center;

            AdaptiveText largeBrandingText = new AdaptiveText();
            largeBrandingText.Text = LocalizedStrings.Get("TileLargeBranding");
            largeBrandingText.HintStyle = AdaptiveTextStyle.CaptionSubtle;
            largeBrandingText.HintAlign = AdaptiveTextAlign.Center;

            largeContent.Children.Add(largeGroup);
            largeContent.Children.Add(largeMessageText);
            largeContent.Children.Add(largeBrandingText);

            content.Visual.TileLarge = new TileBinding();
            content.Visual.TileLarge.Content = largeContent;

            return content.GetXml();
        }

        private static TileBindingContentAdaptive BuildTitleMessageGroup(string title, AdaptiveTextStyle titleStyle, bool wrapTitle, string message, AdaptiveTextStyle messageStyle, int messageMaxLines)
        {
            AdaptiveText titleText = new AdaptiveText();
            titleText.Text = title;
            titleText.HintStyle = titleStyle;
            titleText.HintWrap = wrapTitle;

            AdaptiveText messageText = new AdaptiveText();
            messageText.Text = message;
            messageText.HintStyle = messageStyle;
            messageText.HintWrap = true;
            messageText.HintMaxLines = messageMaxLines;

            AdaptiveSubgroup subgroup = new AdaptiveSubgroup();
            subgroup.Children.Add(titleText);
            subgroup.Children.Add(messageText);

            AdaptiveGroup group = new AdaptiveGroup();
            group.Children.Add(subgroup);

            TileBindingContentAdaptive result = new TileBindingContentAdaptive();
            result.Children.Add(group);
            return result;
        }

        private static TileBranding ParseBranding(string branding)
        {
            switch ((branding ?? "").Trim().ToLowerInvariant())
            {
                case "none":
                    return TileBranding.None;
                case "logo":
                    return TileBranding.Logo;
                case "nameandlogo":
                    return TileBranding.NameAndLogo;
                default:
                    return TileBranding.Name;
            }
        }

        private static string SanitizeText(string text)
        {
            if (string.IsNullOrEmpty(text)) return "";

            StringBuilder sanitized = new StringBuilder(text.Length);
            int i = 0;
            while (i < text.Length)
            {
                char ch = text[i];

                if (char.IsHighSurrogate(ch) && i + 1 < text.Length && char.IsLowSurrogate(text[i + 1]))
                {
                    sanitized.Append(ch);
                    sanitized.Append(text[i + 1]);
                    i += 2;
                    continue;
                }

                if ((ch == '\t') || (ch == '\n') || (ch == '\r') ||
                    (ch >= '\u0020' && ch <= '\uD7FF') ||
                    (ch >= '\uE000' && ch <= '\uFFFD'))
                {
                    sanitized.Append(ch);
                }

                i += 1;
            }

            return sanitized.ToString();
        }

        public static bool BadgeEnabled
        {
            get
            {
                try
                {
                    var stored = Windows.Storage.ApplicationData.Current.LocalSettings.Values[AppConstants.SettingShowTileBadge];
                    if (stored == null) return true;
                    return Convert.ToBoolean(stored);
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"LiveTileService: BadgeEnabled read failed – {ex.GetType().Name}: {ex.Message}");
                    return true;
                }
            }
            set
            {
                try
                {
                    Windows.Storage.ApplicationData.Current.LocalSettings.Values[AppConstants.SettingShowTileBadge] = value;
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"LiveTileService: BadgeEnabled write failed – {ex.GetType().Name}: {ex.Message}");
                }

                if (!value) ClearBadge();
            }
        }

        public static bool TileCleared
        {
            get
            {
                try
                {
                    return Convert.ToBoolean(Windows.Storage.ApplicationData.Current.LocalSettings.Values[AppConstants.SettingLiveTileCleared]);
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"LiveTileService: TileCleared read failed – {ex.GetType().Name}: {ex.Message}");
                    return false;
                }
            }
            set
            {
                try
                {
                    Windows.Storage.ApplicationData.Current.LocalSettings.Values[AppConstants.SettingLiveTileCleared] = value;
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"LiveTileService: TileCleared write failed – {ex.GetType().Name}: {ex.Message}");
                }
            }
        }

        public static void UpdateBadge(int count)
        {
            try
            {
                if (!BadgeEnabled) return;

                if (count <= 0)
                {
                    ClearBadge();
                    return;
                }

                var clampedCount = Math.Min(count, 99);
                var badgeXml = $"<badge value=\"{clampedCount}\"/>";
                XmlDocument badgeDoc = new XmlDocument();
                badgeDoc.LoadXml(badgeXml);

                BadgeNotification badgeNotification = new BadgeNotification(badgeDoc);
                BadgeUpdateManager.CreateBadgeUpdaterForApplication().Update(badgeNotification);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: UpdateBadge failed – {ex.GetType().Name}: {ex.Message}");
            }
        }

        public static void UpdateBadgeGlyph(string glyph)
        {
            if (string.IsNullOrWhiteSpace(glyph)) return;

            try
            {
                if (!BadgeEnabled) return;

                var normalizedGlyph = glyph.Trim();
                if (!IsSupportedBadgeGlyph(normalizedGlyph))
                {
                    Debug.WriteLine($"LiveTileService: UpdateBadgeGlyph skipped unsupported glyph '{normalizedGlyph}'.");
                    return;
                }

                var badgeXml = $"<badge value=\"{normalizedGlyph}\"/>";
                XmlDocument badgeDoc = new XmlDocument();
                badgeDoc.LoadXml(badgeXml);

                BadgeNotification badgeNotification = new BadgeNotification(badgeDoc);
                BadgeUpdateManager.CreateBadgeUpdaterForApplication().Update(badgeNotification);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: UpdateBadgeGlyph failed – {ex.GetType().Name}: {ex.Message}");
            }
        }

        public static bool SendToast(string title, string message)
        {
            try
            {
                var notifier = ToastNotificationManager.CreateToastNotifier();
                if (notifier.Setting != NotificationSetting.Enabled)
                {
                    Debug.WriteLine($"LiveTileService: Toast suppressed – NotificationSetting is {notifier.Setting}. " +
                                    "Enable notifications for this app in Windows Settings > System > Notifications.");
                    return false;
                }

                var toastContent = new ToastContentBuilder()
                    .AddText(SanitizeText(title))
                    .AddText(SanitizeText(message))
                    .GetToastContent();

                ToastNotification toast = new ToastNotification(toastContent.GetXml());
                notifier.Show(toast);
                return true;
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: SendToast failed – {ex.GetType().Name}: {ex.Message}");
                return false;
            }
        }

        public static void ClearTile()
        {
            try
            {
                TileUpdateManager.CreateTileUpdaterForApplication().Clear();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: ClearTile failed – {ex.GetType().Name}: {ex.Message}");
            }
        }

        public static void ClearBadge()
        {
            try
            {
                BadgeUpdateManager.CreateBadgeUpdaterForApplication().Clear();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"LiveTileService: ClearBadge failed – {ex.GetType().Name}: {ex.Message}");
            }
        }

        private static string GetTileMonogram(string primaryText)
        {
            if (string.IsNullOrWhiteSpace(primaryText)) return DefaultMonogram;

            var trimmed = primaryText.Trim();

            var monogram = TextHelper.FirstTextElements(trimmed, 2).ToUpperInvariant();
            return string.IsNullOrEmpty(monogram) ? DefaultMonogram : monogram;
        }

        private static bool IsSupportedBadgeGlyph(string glyph)
        {
            switch (glyph.ToLowerInvariant())
            {
                case "none":
                case "activity":
                case "alarm":
                case "alert":
                case "attention":
                case "available":
                case "away":
                case "busy":
                case "error":
                case "newmessage":
                case "paused":
                case "playing":
                case "unavailable":
                    return true;
                default:
                    return false;
            }
        }
    }

    public class NewsItem
    {
        public string Title { get; set; }
        public string Message { get; set; }
        public string Tag { get; set; }
        public DateTime Timestamp { get; set; }

        public NewsItem(string title, string message, string tag = null)
        {
            this.Title = title;
            this.Message = message;
            this.Tag = tag;
            this.Timestamp = DateTime.Now;
        }
    }
}
